### PR TITLE
feat(web): add pluggable storage backends for RDP file downloads

### PR DESCRIPTION
**Pull request by gabrielbauman**
Right now, `RdpFileTransferProvider` accumulates every download chunk in a `Uint8Array[]` in memory. For a 500 MB file, that means ~1 GB of RAM (the chunks plus the final Blob copy). This PR introduces a `FileStorageBackend` abstraction with two implementations so we can do better when the browser supports it.

### What's new

**`BlobStorageBackend`** - wraps the existing in-memory chunk accumulation behind the new interface. This is the universal fallback and behaves identically to the old code path.

**`OpfsStorageBackend`** - streams download chunks directly to the [Origin Private File System](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API/Origin_private_file_system) via `FileSystemWritableFileStream`. Peak RAM drops from ~2x file size to roughly one chunk (~64 KB). On finalize, `getFile()` returns a lazy `File` handle backed by disk -- no bulk read into memory.

**Automatic detection** - by default, the first download triggers an OPFS probe (create file, open writable, close, delete). If it works, OPFS is used for the session. If not, Blob kicks in silently. Consumers can also force a backend via the new `storageBackend` option on `RdpFileTransferProviderOptions`:

```typescript
// Auto-detect (default)
new RdpFileTransferProvider({ storageBackend: 'auto' })

// Force in-memory
new RdpFileTransferProvider({ storageBackend: 'blob' })

// Bring your own backend
new RdpFileTransferProvider({ storageBackend: myCustomBackend })
```

### OPFS housekeeping

- Each session gets its own subdirectory under `ironrdp-transfers/` in the OPFS root.
- On startup, stale session directories older than 24 hours (from crashed tabs, etc.) are cleaned up in the background.
- On dispose, the session directory is deleted recursively. If that fails, individual files are cleaned up as a fallback.
- Quota exhaustion surfaces as an actionable error on the download, not a silent failure.

### Browser support

The writable stream APIs this depends on reached Baseline across all major browsers in September 2025 (Chrome 86+, Firefox 111+, Safari 17.2+, Edge 86+). Older browsers get the Blob fallback automatically via the probe -- no breakage.

### API surface

All new types are re-exported from the package entry point:

- `FileStorageBackend` (interface)
- `FileWriteHandle` (interface)
- `StorageBackendPreference` (type: `'auto' | 'blob'`)
- `BlobStorageBackend` (class)
- `OpfsStorageBackend` (class)
- `detectStorageBackend` (function)

The only change to the existing public API is the new optional `storageBackend` field on `RdpFileTransferProviderOptions`. No breaking changes.

## Testing

97 tests covering both backends, detection/fallback logic, download flow integration, sanitization of adversarial file names, dispose/abort lifecycle edge cases, and error paths (quota, timeouts, init failures). No new runtime dependencies.

### CODE REVIEW

File: web-client/iron-remote-desktop-rdp/src/storage/OpfsStorageBackend.ts [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    // Ensure we always have a non-empty name.
+    if (safe.length === 0) {
+        safe = 'unnamed';
+    }
+
+    // OPFS entry names are typically limited to 255 bytes.  Truncate to
+    // leave room for the sequence prefix added by createWriteHandle.
+    if (safe.length > 200) {
+        safe = safe.slice(0, 200);
+    }
```
[the remark this answers is not in the document]
Comment (gabrielbauman) [reply web-client/iron-remote-desktop-rdp/src/storage/OpfsStorageBackend.ts]: Fixed -- sanitizeOpfsName now uses TextEncoder to measure and enforce a 200 UTF-8 byte limit instead of character count.

File: web-client/iron-remote-desktop-rdp/src/RdpFileTransferProvider.ts [hunk 2]
```diff
@@ -1352,40 +1445,160 @@ export class RdpFileTransferProvider {
...
-            state.chunks.push(response.data);
-            state.bytesReceived += response.data.length;
-
-            // Validate that received data doesn't grossly exceed expected size
-            if (state.bytesReceived > state.expectedSize * 2) {
-                this.activeDownloads.delete(response.streamId);
-                state.chunks = [];
+            // This is a DATA response -- write the chunk to the storage backend.
+            void this.handleDataChunk(state, response.data);
+        }
```
[the remark this answers is not in the document]
Comment (gabrielbauman) [reply web-client/iron-remote-desktop-rdp/src/RdpFileTransferProvider.ts:1461]: Not applicable here. The RDP file transfer protocol is strictly request-response: requestNextChunk is only called after write() completes inside handleDataChunk (line 1596), so the server cannot send a second DATA response until we ask for one. There is no concurrent dispatch path into handleDataChunk for the same stream.

File: web-client/iron-remote-desktop-rdp/src/storage/detect.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            // Probe failed: the OPFS API exists but is not functional
+            // (e.g., createWritable() throws in some browser modes).
+            console.warn('OPFS probe failed (createWritable not functional), falling back to blob storage');
+        } catch (error) {
+            // getDirectory() itself threw (e.g., SecurityError in some
+            // private browsing modes).  Fall through to Blob.
+            console.warn('OPFS unavailable, falling back to blob storage:', error);
+        }
```
[the remark this answers is not in the document]
Comment (gabrielbauman) [reply web-client/iron-remote-desktop-rdp/src/storage/detect.ts:58]: Fair point -- downgraded both to console.debug. In private browsing OPFS is always unavailable and warn would fire every session.

File: web-client/iron-remote-desktop-rdp/src/storage/FileStorageBackend.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ * storage concerns to the active backend, keeping download orchestration
+ * logic storage-agnostic.
+ *
+ * Three backends are planned:
+ *
+ * | Backend | Buffering          | Peak RAM      | Browser support        |
+ * |---------|--------------------|---------------|------------------------|
+ * | Blob    | In-memory array    | ~2x file size | Universal              |
+ * | OPFS    | Origin Private FS  | ~chunk size   | Baseline (March 2023)  |
+ * | FSAPI   | User-chosen file   | ~chunk size   | Chromium-only (future) |
```
[the remark this answers is not in the document]
Comment (gabrielbauman) [reply web-client/iron-remote-desktop-rdp/src/storage/FileStorageBackend.ts]: Fixed -- table now says September 2025 to match OpfsStorageBackend docs.

**Review by CBenoit — COMMENTED** [review]
Heads up: I intended to review this sooner, but ran out of time, I'm sorry! I'll be away until next week, I'll get through it then =)

**Comment by ymarcus93** [comment]
Hey @CBenoit, gentle bump, do you mind taking a look at this PR?

**Comment by CBenoit** [comment]
@ymarcus93 Ping acknowledged. I’ll go through important PRs, including this one by Friday.

**Review by CBenoit — APPROVED** [review]
I’m sorry for the delay, I finally went through this! It’s looking pretty good to me 👍

@ymarcus93 are you in position of pushing a commit to fix the conflict, or should I handle that?

**Comment by ymarcus93** [comment]
> I’m sorry for the delay, I finally went through this! It’s looking pretty good to me 👍
> 
> Yuval Marcus (@ymarcus93) are you in position of pushing a commit to fix the conflict, or should I handle that?

Thank you!

I can handle the conflict, no worries. Will wait for #1381 to be merged first just in case

**Comment by ymarcus93** [comment]
Hmm actually, I don't think I can since I don't own the source branch. I could make a duplicate PR thats cherry-picked and rebased with conflicts resolved

**Comment by CBenoit** [comment]
Thank you for proposing; I think it will be fine, @gabrielbauman enabled maintainers to push to his branch, I’ll try that.

**Comment by CBenoit** [comment]
Looks like I was able to address the conflicts on my side 🙂 

Thank you @gabrielbauman!
